lib/src/klio/message/pubsub_message_manager.py

```diff
@@ -127,6 +127,21 @@ def heartbeat(self, message):
                 f"Job is still processing {message.kmsg_id}..."
             )
             time.sleep(self.heartbeat_sleep)
+            # Sometimes the heartbeat worker still continues in the `while`
+            # loop even though the message event flag has been set (not sure
+            # how this can happen...).
+            # So we're just checking to see if the message we're looping
+            # over still exists in ENTITY_ID_TO_ACK_ID; and if not, to
+            # go ahead and remove it. The `remove` method should be
+            # idempotent.
+            if message.kmsg_id not in ENTITY_ID_TO_ACK_ID:
+                self.hrt_logger.debug(
+                    f"No record of message {message.kmsg_id}. Assuming the "
+                    "pipeline is done processing it and will try to ack the "
+                    "message..."
+                )
+                self.remove(message)
```

Comment:
If we're adding this in, is `message.event` even needed anymore?  Would it make more sense to just have `while message.kmsg_id in ENTITY_ID_TO_ACK_ID`?

Reply:
(responding to both this comment and your overall PR comment)

Good questions. So, I reached for thread event objects as a thread-safe means to control/signal (not `signal` signal) the two worker threads from the main thread (this [SO comment](https://stackoverflow.com/questions/18485098/python-threading-with-event-object#comment119832690_18485226) goes into detail). I also like using thread event objects because, at least to me, it signals (to the code reader) intent that, "this is a separate thread, and is dependent on another thread's actions." Or, at least that was _my_ intent. 

I'm happy to change it to be based on a dict lookup. I think I'd need to do a `while True`, and add a thread lock since the `ENTITY_ID_TO_ACK_ID` is meant to be modified (below there's `ENTITY_ID_TO_ACK_ID.pop(psk_msg.kmsg_id, None)`) and looping over a dict that may change in size can lead to unknown state.

Reply:
Latest changes look good to me, using the lock makes sense too.  I think what I said with events was more of a guess and depends a lot on Python internals around multi-threading and caching.

On a similar note, how does Python handle accessing the same dict across threads, could there be a caching issue here too?  For reference, in Java you need to declare variables as `volatile` if you plan to write on one thread and read from another, since otherwise the write isn't always propagated across threads.  I have no idea if the same issues arise in Python.